File: src/utils/parser.ts
import { AgentResponse } from '../verifiers/BaseVerifier.js';

export class ZGuardParser {
  /**
   * Parses raw text output from an agent to extract factual claims and citation attachments.
   * Supports two formats:
   * 1. XML-style: <claim citation="doc_01">Acme revenue was $15.4M</claim>
   * 2. Inline brackets: Acme revenue was $15.4M [doc_01]
   */
  static parseClaims(text: string): AgentResponse[] {
    const claims: AgentResponse[] = [];

    // 1. XML tag parser: <claim citation="id">claim text</claim>
    const xmlRegex = /<claim\s+citation="([^"]+)"\s*>([\s\S]*?)<\/claim>/gi;
    let xmlMatch;
    while ((xmlMatch = xmlRegex.exec(text)) !== null) {
      claims.push({
        citationId: xmlMatch[1].trim(),
        factualClaim: xmlMatch[2].trim()
      });
    }

    // If XML matches are found, we prioritize them
    if (claims.length > 0) {
      return claims;
    }

    // 2. Bracket-style parser: matches sentences ending with [id] or [id:lines]
    // Matches: "Any text before the bracket [doc_01]" or "Any text [doc_02:10-15]"
    // Splitting sentences based on period/newline, then matching brackets
    const sentences = text.split(/(?<=[.!?])\s+|\n+/);
    const bracketRegex = /([\s\S]+?)\s*\[([a-zA-Z0-9_-]+)(?::[0-9a-zA-Z_-]+)?\][.!?]?\s*$/i;

    for (const sentence of sentences) {
      const match = sentence.trim().match(bracketRegex);
      if (match) {
        claims.push({
          citationId: match[2].trim(),
          factualClaim: match[1].trim()
        });
      }
    }

    return claims;
  }

  /**
   * Parses a tool call from the text if it is represented in a tag structure:
   * <tool name="toolName">{ "param": "value" }</tool>
   */
  static parseToolCall(text: string): { toolCall?: string; toolArgs?: Record<string, any> } {
    const toolRegex = /<tool\s+name="([^"]+)"\s*>([\s\S]*?)<\/tool>/i;
    const match = text.match(toolRegex);

    if (match) {
      const toolCall = match[1].trim();
      const rawArgs = match[2].trim();
      try {
        const toolArgs = JSON.parse(rawArgs);
        return { toolCall, toolArgs };
      } catch (err) {
        return {
          toolCall,
          toolArgs: { _raw: rawArgs, _error: 'JSON parsing failed' }
        };
      }
    }

    return {};
  }
}

File: zguard-python/zguard/parser.py
import re
import json
from typing import List, Dict, Any
from .types import AgentResponse

class ZGuardParser:
    """
    Parses raw text output from an agent to extract factual claims and citation attachments.
    Supports two formats:
    1. XML-style: <claim citation="doc_01">Acme revenue was $15.4M</claim>
    2. Inline brackets: Acme revenue was $15.4M [doc_01]
    """

    @staticmethod
    def parse_claims(text: str) -> List[AgentResponse]:
        claims: List[AgentResponse] = []

        # 1. XML tag parser: <claim citation="id">claim text</claim>
        xml_regex = re.compile(r'<claim\s+citation="([^"]+)"\s*>([\s\S]*?)</claim>', re.IGNORECASE)
        for match in xml_regex.finditer(text):
            claims.append({
                "citationId": match.group(1).strip(),
                "factualClaim": match.group(2).strip()
            })

        if claims:
            return claims

        # 2. Bracket-style parser: matches sentences ending with [id] or [id:lines]
        sentences = re.split(r'(?<=[.!?])\s+|\n+', text)
        bracket_regex = re.compile(r'([\s\S]+?)\s*\[([a-zA-Z0-9_-]+)(?::[0-9a-zA-Z_-]+)?\][.!?]?\s*$', re.IGNORECASE)

        for sentence in sentences:
            sentence_str = sentence.strip()
            if not sentence_str:
                continue
            match = bracket_regex.match(sentence_str)
            if match:
                claims.append({
                    "citationId": match.group(2).strip(),
                    "factualClaim": match.group(1).strip()
                })

        return claims

    @staticmethod
    def parse_tool_call(text: str) -> Dict[str, Any]:
        """
        Parses a tool call from the text if it is represented in a tag structure:
        <tool name="toolName">{ "param": "value" }</tool>
        """
        tool_regex = re.compile(r'<tool\s+name="([^"]+)"\s*>([\s\S]*?)</tool>', re.IGNORECASE)
        match = tool_regex.search(text)

        if match:
            tool_call = match.group(1).strip()
            raw_args = match.group(2).strip()
            try:
                tool_args = json.loads(raw_args)
                return {"toolCall": tool_call, "toolArgs": tool_args}
            except Exception:
                return {
                    "toolCall": tool_call,
                    "toolArgs": {"_raw": raw_args, "_error": "JSON parsing failed"}
                }

        return {}
